projects/shadow-4.16.0/man/man8/sulogin.8
'\" t
.\"     Title: sulogin
.\"    Author: Julianne Frances Haugh
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 06/18/2024
.\"    Manual: System Management Commands
.\"    Source: shadow-utils 4.16.0
.\"  Language: English
.\"
.TH "SULOGIN" "8" "06/18/2024" "shadow\-utils 4\&.16\&.0" "System Management Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
sulogin \- Single\-user login
.SH "SYNTAX"
.PP
\fBsulogin\fR
[\fItty\-device\fR]
.SH "DESCRIPTION"
.PP
The
\fBsulogin\fR
command is invoked by
\fBinit\fR
prior to allowing the user access to the system when in single user mode\&. This feature may only be available on certain systems where
\fBinit\fR
has been modified accordingly, or where the
/etc/inittab
has an entry for a single user login\&.
.PP
The user is prompted
.PP
Type control\-d to proceed with normal startup,
(or give root password for system maintenance):
.PP
Input and output will be performed with the standard file descriptors unless the optional device name argument is provided\&.
.PP
If the user enters the correct root password, a login session is initiated\&. When
\fIEOF\fR
is pressed instead, the system enters multi\-user mode\&.
.PP
After the user exits the single\-user shell, or presses
\fIEOF\fR, the system begins the initialization process required to enter multi\-user mode\&.
.SH "CAVEATS"
.PP
This command can only be used if
\fBinit\fR
has been modified to call
\fBsulogin\fR
instead of
/bin/sh, or if the user has set the
\fIinittab\fR
to support a single user login\&. For example, the line:
.PP
co:s:respawn:/etc/sulogin /dev/console
.PP
should execute the sulogin command in single user mode\&.
.PP
As complete an environment as possible is created\&. However, various devices may be unmounted or uninitialized and many of the user commands may be unavailable or nonfunctional as a result\&.
.SH "CONFIGURATION"
.PP
The following configuration variables in
/etc/login\&.defs
change the behavior of this tool:
.PP
\fBENV_HZ\fR (string)
.RS 4
If set, it will be used to define the HZ environment variable when a user login\&. The value must be preceded by
\fIHZ=\fR\&. A common value on Linux is
\fIHZ=100\fR\&.
.RE
.PP
\fBENV_TZ\fR (string)
.RS 4
If set, it will be used to define the TZ environment variable when a user login\&. The value can be the name of a timezone preceded by
\fITZ=\fR
(for example
\fITZ=CST6CDT\fR), or the full path to the file containing the timezone specification (for example
/etc/tzname)\&.
.sp
If a full path is specified but the file does not exist or cannot be read, the default is to use
\fITZ=CST6CDT\fR\&.
.RE
.SH "FILES"
.PP
/etc/passwd
.RS 4
User account information\&.
.RE
.PP
/etc/shadow
.RS 4
Secure user account information\&.
.RE
.SH "SEE ALSO"
.PP
\fBlogin\fR(1),
\fBsh\fR(1),
\fBinit\fR(8)\&.


projects/shadow-4.16.0/src/sulogin.c
/*
 * SPDX-FileCopyrightText: 1989 - 1994, Julianne Frances Haugh
 * SPDX-FileCopyrightText: 1996 - 2000, Marek Michałkiewicz
 * SPDX-FileCopyrightText: 2002 - 2006, Tomasz Kłoczko
 * SPDX-FileCopyrightText: 2007 - 2010, Nicolas François
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include <config.h>

#ident "$Id$"

#include <fcntl.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <sys/ioctl.h>
#include <sys/types.h>

#include "agetpass.h"
#include "alloc.h"
#include "attr.h"
#include "defines.h"
#include "getdef.h"
#include "prototypes.h"
#include "pwauth.h"
/*@-exitarg@*/
#include "exitcodes.h"
#include "shadowlog.h"


/*
 * Global variables
 */
static const char Prog[] = "sulogin";


extern char **newenvp;

#ifndef	ALARM
#define	ALARM	60
#endif


static void catch_signals (int);
static int pw_entry(const char *name, struct passwd *pwent);


static void catch_signals (MAYBE_UNUSED int sig)
{
	_exit (1);
}


int
main(int argc, char *argv[])
{
	int            err = 0;
	char           **envp = environ;
	TERMIO         termio;
	struct passwd  pwent = {};
	bool           done;
#ifndef USE_PAM
	const char     *env;
#endif


	tcgetattr (0, &termio);
	termio.c_iflag |= (ICRNL | IXON);
	termio.c_oflag |= (CREAD);
	termio.c_lflag |= (ECHO | ECHOE | ECHOK | ICANON | ISIG);
	tcsetattr (0, TCSANOW, &termio);

	log_set_progname(Prog);
	log_set_logfd(stderr);
	(void) setlocale (LC_ALL, "");
	(void) bindtextdomain (PACKAGE, LOCALEDIR);
	(void) textdomain (PACKAGE);

	initenv();
	if (argc > 1) {
		close(0);
		close(1);
		close(2);

		if (open(argv[1], O_RDWR) == -1)
			exit(1);
		dup(0);
		dup(0);
	}
	if (access (PASSWD_FILE, F_OK) == -1) {	/* must be a password file! */
		(void) puts (_("No password file"));
		exit (1);
	}
#if !defined(DEBUG) && defined(SULOGIN_ONLY_INIT)
	if (getppid () != 1) {	/* parent must be INIT */
		exit (1);
	}
#endif
	if ((isatty (0) == 0) || (isatty (1) == 0) || (isatty (2) == 0)) {
		exit (1);	/* must be a terminal */
	}
	/* If we were init, we need to start a new session */
	if (getppid() == 1) {
		setsid();
		if (ioctl(0, TIOCSCTTY, 1) != 0) {
			(void) fputs (_("TIOCSCTTY failed"), stderr);
		}
	}
	while (NULL != *envp) {		/* add inherited environment, */
		addenv (*envp, NULL);	/* some variables change later */
		envp++;
	}

#ifndef USE_PAM
	env = getdef_str ("ENV_TZ");
	if (NULL != env) {
		addenv (('/' == *env) ? tz (env) : env, NULL);
	}
	env = getdef_str ("ENV_HZ");
	if (NULL != env) {
		addenv (env, NULL);	/* set the default $HZ, if one */
	}
#endif				/* !USE_PAM */

	(void) signal (SIGALRM, catch_signals);	/* exit if the timer expires */
	(void) alarm (ALARM);		/* only wait so long ... */

	do {			/* repeatedly get login/password pairs */
		char *pass;
		if (pw_entry("root", &pwent) == -1) {	/* get entry from password file */
			/*
			 * Fail secure
			 */
			(void) puts(_("No password entry for 'root'"));
			exit(1);
		}

		/*
		 * Here we prompt for the root password, or if no password
		 * is given we just exit.
		 */

		/* get a password for root */
		pass = agetpass (_(
"\n"
"Type control-d to proceed with normal startup,\n"
"(or give root password for system maintenance):"));
		/*
		 * XXX - can't enter single user mode if root password is
		 * empty.  I think this doesn't happen very often :-). But
		 * it will work with standard getpass() (no NULL on EOF).
		 * --marekm
		 */
		if ((NULL == pass) || ('\0' == *pass)) {
			erase_pass (pass);
			(void) puts ("");
#ifdef	TELINIT
			execl (PATH_TELINIT, "telinit", RUNLEVEL, (char *) NULL);
#endif
			exit (0);
		}

		done = valid(pass, &pwent);
		erase_pass (pass);

		if (!done) {	/* check encrypted passwords ... */
			/* ... encrypted passwords did not match */
			sleep (2);
			(void) puts (_("Login incorrect"));
		}
	} while (!done);
	(void) alarm (0);
	(void) signal (SIGALRM, SIG_DFL);
	environ = newenvp;	/* make new environment active */

	(void) puts (_("Entering System Maintenance Mode"));

	/* exec the shell finally. */
	err = shell (pwent.pw_shell, NULL, environ);

	return ((err == ENOENT) ? E_CMD_NOTFOUND : E_CMD_NOEXEC);
}


static int
pw_entry(const char *name, struct passwd *pwent)
{
	struct spwd    *spwd;
	struct passwd  *passwd;

	if (!(passwd = getpwnam(name)))  /* local, no need for xgetpwnam */
		return -1;

	free(pwent->pw_name);
	pwent->pw_name = xstrdup(passwd->pw_name);
	pwent->pw_uid = passwd->pw_uid;
	pwent->pw_gid = passwd->pw_gid;
	free(pwent->pw_gecos);
	pwent->pw_gecos = xstrdup(passwd->pw_gecos);
	free(pwent->pw_dir);
	pwent->pw_dir = xstrdup(passwd->pw_dir);
	free(pwent->pw_shell);
	pwent->pw_shell = xstrdup(passwd->pw_shell);
#if !defined(AUTOSHADOW)
	/* local, no need for xgetspnam */
	if ((spwd = getspnam(name))) {
		free(pwent->pw_passwd);
		pwent->pw_passwd = xstrdup(spwd->sp_pwdp);
		return 0;
	}
#endif
	free(pwent->pw_passwd);
	pwent->pw_passwd = xstrdup(passwd->pw_passwd);
	return 0;
}
